lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,348 @@
+  getBuilder().setInsertionPoint(entryBlock);
+  getBuilder().createBranch(RegularLocation::getAutoGeneratedLocation(),
+                            getOpBasicBlock(cloneFromBlock));
+}
+
+/// Update address uses of the opaque type archetype with the concrete type.
+/// This is neccessary for apply instructions.
+void OpaqueSpecializerCloner::insertOpaqueToConcreteAddressCasts(
+    SILInstruction *orig, SILInstruction *cloned) {
+
+  // Replace apply operands.
+  if (auto apply = ApplySite::isa(cloned)) {
+    SavedInsertionPointRAII restore(getBuilder());
+    getBuilder().setInsertionPoint(apply.getInstruction());
+    auto substConv = apply.getSubstCalleeConv();
+    unsigned idx = 0;
+    for (auto &opd : apply.getArgumentOperands()) {
+      auto argConv = apply.getArgumentConvention(opd);
+      auto argIdx = apply.getCalleeArgIndex(opd);
+      auto argType = substConv.getSILArgumentType(argIdx);
+      if (argType.getASTType() != opd.get()->getType().getASTType()) {
+        if (argConv.isIndirectConvention()) {
+          auto cast = getBuilder().createUncheckedAddrCast(apply.getLoc(),
+                                                           opd.get(), argType);
+          opd.set(cast);
+        } else {
+          auto cast = getBuilder().createUncheckedRefCast(apply.getLoc(),
+                                                          opd.get(), argType);
+          opd.set(cast);
+        }
+      }
+      ++idx;
+    }
+    return;
+  }
+}
+
+namespace {
+class OpaqueArchetypeSpecializer : public SILFunctionTransform {
+  void run() override {
+    auto *currentModule = getFunction()->getModule().getSwiftModule();
+
+    auto opaqueArchetypeWouldChange = [=](CanType ty) -> bool {
+      if (!ty->hasOpaqueArchetype())
+        return false;
+      ReplaceOpaqueTypesWithUnderlyingTypes replacer(currentModule);
+      return ty.subst(replacer, replacer,
+                      SubstFlags::SubstituteOpaqueArchetypes |
+                          SubstFlags::AllowLoweredTypes)
+                 ->getCanonicalType() != ty;
+    };
+
+    // Look for opaque type archetypes.
+    bool foundOpaqueArchetype = false;
+    for (auto &BB : *getFunction()) {
+      for (auto &inst : BB) {
+        for (auto &opd : inst.getAllOperands()) {
+          if (!opaqueArchetypeWouldChange(opd.get()->getType().getASTType()))
+            continue;
+          foundOpaqueArchetype = true;
+          break;
+        }
+        if (foundOpaqueArchetype)
+          break;
+        auto *allocStack = dyn_cast<AllocStackInst>(&inst);
+        if (!allocStack || !opaqueArchetypeWouldChange(
+                               allocStack->getElementType().getASTType()))
+          continue;
+        foundOpaqueArchetype = true;
+        break;
+      }
+    }
+
+    if (foundOpaqueArchetype) {
+      SubstitutionMap subsMap = getFunction()->getForwardingSubstitutionMap();
+      OpaqueSpecializerCloner s(subsMap, *getFunction());
```

Comment:
the subsMap should not be used at all, I think you can try removing it. It's just burning CPU cycles to substitute using forwarding substitutions. Maybe the opaque cloner should directly inherit from SILCloner and not TypeSubstCloner?